### ISSUE TITLE
Feature request: forwards autodifferentiation for gradient computation

### ISSUE THREAD

**Issue by ojwoodford** [tensorflow/tensorflow#19361]
Does TensorFlow use forwards instead of reverse (i.e. backprop) autodiff where appropriate? 

For graphs with fewer inputs (layer input values plus model parameters) than outputs (as  occurs in convolutional upsampling architectures now popular in GANs and auto-encoders), I believe forwards auto-differentiation will be a faster way to compute gradients. The Jacobian matrices get progressively wider as you go from right to left (up the chain of functions), so if you start multiplying from the right (forwards), rather than from the left (backprop) you reduce the number of operations.

But maybe I'm missing something.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by ojwoodford** [comment]
- [ ] Have I written custom code
- [ ] OS Platform and Distribution
- [ ] TensorFlow installed from
- [ ] TensorFlow version
- [ ] Bazel version
- [ ] CUDA/cuDNN version
- [ ] GPU model and memory
- [ ] Exact command to reproduce

**Comment by ojwoodford** [comment]
Great. Thanks!
